### PR TITLE
Track shard snapshot progress during node shutdown

### PR DESCRIPTION
**Pull request by DiannaHohensee**
Track shard snapshot progress during shutdown to identify any
bottlenecks that cause slowness that can ultimately block shard
re-allocation.

-------------
TODO: note to self, include human readable time since shutdown received, time it took to set PAUSING. And the time it took to finish stopping all the shard snapshots, to completely unblock shard re-allocation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/ReferenceDocs.java [hunk 1]
```diff
@@ -80,6 +80,7 @@ public enum ReferenceDocs {
     LUCENE_MAX_DOCS_LIMIT,
     MAX_SHARDS_PER_NODE,
     FLOOD_STAGE_WATERMARK,
+    STATELESS_EMPTY,
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/common/ReferenceDocs.java:83]: going to pull this out into a separate patch

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 2]
```diff
@@ -689,4 +713,17 @@ public void onFailure(Exception e) {
             )
         );
     }
+
+    /**
+     * TODO: should I use Stage or some other state? It seems the most precise for having PAUSING etc, but look around.
+     */
+    public Map<Stage, Long> getCountOfInProgressShardSnapshotsByShardSnapshotStage() {
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:720]: Changed to actively tracking shard snapshots, rather than counting them on demand. Keeping this around for discussion, but plan to remove it.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            shutdownFinishedSignallingPausingMillis,
+            numberOfShardSnapshotsInProgressOnDataNode.get(),
+            shardSnapshotRequests.size(),
+            doneCount.get(),
+            failureCount.get(),
+            abortedCount.get(),
+            pausedCount.get()
+        );
+
+        HotThreads.logLocalHotThreads(logger, Level.INFO, "slow snapshot pausing", ReferenceDocs.STATELESS_EMPTY);
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:219]: On further thought, I don't think this belongs in the tracker. I'll pull it out and have it as a standalone someplace for dumping hot threads during slow shutdown, unassociated with snapshots in particular.

**Comment by DiannaHohensee** [comment]
@DaveCTurner Could you take a look at this to agree on direction before I go work on testing it?

This is the tracking I had in mind. If we have any other ideas, or get some initial results and want more specifics, I figure that will be follow up work/additions. We might want to know more about the ShardSnapshotTaskRunner and network calls, for example, however that may be attainable.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 4]
```diff
@@ -130,10 +133,29 @@ protected void doClose() {
...
+            final var previousSnapshots = SnapshotsInProgress.get(event.previousState());
+            if (previousSnapshots.equals(currentSnapshots) == false) {
                 final var localNodeId = clusterService.localNode().getId();
+
+                {
+                    // Track when this node enters and leaves shutdown mode because we pause shard snapshots for shutdown.
+                    // The snapshotShutdownProgressTracker will report (via logging) on the progress shard snapshots make
+                    // towards either completing (successfully or otherwise) or pausing.
+                    final var previouslyInShutdownMode = previousSnapshots.isNodeIdForRemoval(localNodeId);
+                    final var currentlyInShutdownMode = previousSnapshots.isNodeIdForRemoval(localNodeId);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:145]: typo I think?

```suggestion
                    final var currentlyInShutdownMode = currentSnapshots.isNodeIdForRemoval(localNodeId);
```

**Review by DaveCTurner — COMMENTED** [review]
Seems reasonable to me. Not sure we want this at INFO in the main ES codebase, but I'm also not sure we don't want that.

**Comment by idegtiarenko** [comment]
Corresponding issue is in review, however this change is in draft status.
Is it ready for the review?
